Sources/Fuzzilli/Core/CodeTemplate.swift

```diff
@@ -0,0 +1,114 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+public class CodeTemplate {
+    public let name: String
+
+    public var stats = ProgramGeneratorStats()
+
+    private let f: (ProgramBuilder) -> ()
+
+    init(_ name: String, _ f: @escaping (_: ProgramBuilder) -> ()) {
+        self.name = name
+        self.f = f
+    }
+
+    func generate(in b: ProgramBuilder) {
+        f(b)
+    }
+
+    /// Generate an array of random function signatures
+    static func generateRandomFunctionSignatures(forFuzzer fuzzer: Fuzzer, n: Int) -> [FunctionSignature] {
+        var signatures = [FunctionSignature]()
+        for _ in 0..<n {
+            signatures.append(CodeTemplate.generateSignature(forFuzzer: fuzzer, n: Int.random(in: 0...3)))
+        }
+        return signatures
+    }
+
+    /// This function generates and sets property types for the global properties
+    static func generateRandomPropertyTypes(forBuilder b: ProgramBuilder) {
+        // generate types for half of the available property names.
+        for i in 0..<b.fuzzer.environment.customPropertyNames.count/2 {
+            // if we pass "e" (last customPropertyName) into generateType the filter might fail and we
```

Comment:
I think `generateType` should just be able to handle the case that the filter returns an empty array. Then this workaround isn't necessary.
IIUC, the current approach for preventing cycles works by ensuring that a property X can only be an object with properties whose names are all > X. Then it's guaranteed that you will never get a cycle. However, that implies that the property "e" (the one with the "highest" value) cannot be an object with any other properties.